drivers/ethernet/eth_sam_gmac.c

```diff
@@ -2109,239 +2036,210 @@ static const struct ethernet_api eth_api = {
-		    queue3_isr, DEVICE_DT_INST_GET(0), 0);
-	irq_enable(DT_INST_IRQ_BY_NAME(0, q3, irq));
+#define DEFN_IRQ_CONFIG_3(n) DEFN_IRQ_CONFIG(n, 3, q3)
+#else
+#define DEFN_IRQ_CONFIG_3(n)
 #endif
-
 #if GMAC_ACTIVE_PRIORITY_QUEUE_NUM >= 4
-	IRQ_CONNECT(DT_INST_IRQ_BY_NAME(0, q4, irq),
-		    DT_INST_IRQ_BY_NAME(0, q4, priority),
-		    queue4_isr, DEVICE_DT_INST_GET(0), 0);
-	irq_enable(DT_INST_IRQ_BY_NAME(0, q4, irq));
+#define DEFN_IRQ_CONFIG_4(n) DEFN_IRQ_CONFIG(n, 4, q4)
+#else
+#define DEFN_IRQ_CONFIG_4(n)
 #endif
-
 #if GMAC_ACTIVE_PRIORITY_QUEUE_NUM >= 5
-	IRQ_CONNECT(DT_INST_IRQ_BY_NAME(0, q5, irq),
-		    DT_INST_IRQ_BY_NAME(0, q5, priority),
-		    queue5_isr, DEVICE_DT_INST_GET(0), 0);
-	irq_enable(DT_INST_IRQ_BY_NAME(0, q5, irq));
-#endif
-}
+#define DEFN_IRQ_CONFIG_5(n) DEFN_IRQ_CONFIG(n, 5, q5)
+#else
+#define DEFN_IRQ_CONFIG_5(n)
+#endif
+#define SAM_GMAC_IRQ_CONFIG_DEFN(n)							\
+		static void eth##n##_irq_config(void)					\
+		{									\
+			DEFN_IRQ_CONFIG_0(n)						\
+			DEFN_IRQ_CONFIG_1(n)						\
+			DEFN_IRQ_CONFIG_2(n)						\
+			DEFN_IRQ_CONFIG_3(n)						\
+			DEFN_IRQ_CONFIG_4(n)						\
+			DEFN_IRQ_CONFIG_5(n)						\
+		}
 
-PINCTRL_DT_INST_DEFINE(0);
+#define SAM_GMAC_PINCTRL_DEFN(n) PINCTRL_DT_INST_DEFINE(n);
 
-static const struct eth_sam_dev_cfg eth0_config = {
-	.regs = (Gmac *)DT_INST_REG_ADDR(0),
-	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
 #ifdef CONFIG_SOC_FAMILY_ATMEL_SAM
-	.clock_cfg = SAM_DT_INST_CLOCK_PMC_CFG(0),
-#endif
-	.config_func = eth0_irq_config,
-	.phy_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, phy_handle))
-};
-
-static struct eth_sam_dev_data eth0_data = {
-#if NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))
-	.mac_addr = DT_INST_PROP(0, local_mac_address),
-#endif
-	.queue_list = {
-		{
-			.que_idx = GMAC_QUE_0,
-			.rx_desc_list = {
-				.buf = rx_desc_que0,
-				.len = ARRAY_SIZE(rx_desc_que0),
-			},
-			.tx_desc_list = {
-				.buf = tx_desc_que0,
-				.len = ARRAY_SIZE(tx_desc_que0),
-			},
-			.rx_frag_list = rx_frag_list_que0,
-#if GMAC_MULTIPLE_TX_PACKETS == 1
-			.tx_frag_list = {
-				.buf = (uint32_t *)tx_frag_list_que0,
-				.len = ARRAY_SIZE(tx_frag_list_que0),
-			},
-#if defined(CONFIG_PTP_CLOCK_SAM_GMAC)
-			.tx_frames = {
-				.buf = (uint32_t *)tx_frame_list_que0,
-				.len = ARRAY_SIZE(tx_frame_list_que0),
-			},
-#endif
+#define CFG_CLK_DEFN(n) .clock_cfg = SAM_DT_INST_CLOCK_PMC_CFG(n),
+#else
+#define CFG_CLK_DEFN(n)
+#endif
+
+#define SAM_GMAC_MAC_EEPROM_I2C(n)							\
+		COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), mac_eeprom),		\
+				(I2C_DT_SPEC_GET(DT_INST_PHANDLE(n, mac_eeprom))),	\
+				({}))
+
+#define SAM_GMAC_MAC_ADDR_SOURCE(n)							\
+		COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), mac_eeprom),		\
+				(MAC_ADDR_SOURCE_EEPROM),				\
+		(COND_CODE_1(DT_INST_PROP(n, zephyr_random_mac_address),		\
+				(MAC_ADDR_SOURCE_RANDOM),				\
+		(COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), local_mac_address),	\
+				(MAC_ADDR_SOURCE_LOCAL),				\
+				(MAC_ADDR_SOURCE_INVALID))))))
```

Comment:
no need to check for local address, also a invalid mac address should not stop the init as the user can change it later via the api.